File: nekoshlink/shlink-security/src/main/java/org/nekosoft/shlink/sec/cli/AuthenticationExecutionStrategy.java
package org.nekosoft.shlink.sec.cli;

import org.nekosoft.shlink.sec.ApiKeyAuthenticationToken;
import org.springframework.boot.autoconfigure.condition.ConditionalOnNotWebApplication;
import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.server.resource.BearerTokenAuthenticationToken;
import org.springframework.stereotype.Component;
import picocli.CommandLine.*;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.util.HashSet;

@Component
@ConditionalOnNotWebApplication
public class AuthenticationExecutionStrategy implements IExecutionStrategy {

    private AuthenticationManager authManager;

    public AuthenticationExecutionStrategy(AuthenticationManager authManager) {
        this.authManager = authManager;
    }

    @Override
    public int execute(ParseResult parseResult) throws ExecutionException, ParameterException {
        String username = parseResult.matchedOption("--usr") == null
                ? null : parseResult.matchedOption("--usr").getValue();
        String password = parseResult.matchedOption("--pwd") == null
                ? null : parseResult.matchedOption("--pwd").getValue();
        String apiKey = parseResult.matchedOption("--api-key") == null
                ? null : parseResult.matchedOption("--api-key").getValue();
        File accessTokenFile = parseResult.matchedOption("--access-token-file") == null
                ? null : parseResult.matchedOption("--api-key").getValue();
        Authentication authentication;
        if (accessTokenFile != null && accessTokenFile.canRead()) {
            try {
                authentication = new BearerTokenAuthenticationToken(Files.readString(accessTokenFile.toPath()));
            } catch (IOException e) {
                throw new ExecutionException(parseResult.commandSpec().commandLine(), e.getMessage());
            }
        } else if ((username != null && !username.isBlank()) && (password != null && !password.isBlank())) {
            authentication = new UsernamePasswordAuthenticationToken(username, password);
        } else if (username != null && !username.isBlank()) {
            authentication = new ApiKeyAuthenticationToken(apiKey);
        } else {
            HashSet<GrantedAuthority> authorities = new HashSet<>();
            authorities.add(new SimpleGrantedAuthority("ROLE_Anyone"));
            authentication = new AnonymousAuthenticationToken("NekoShlink", "anonymous", authorities);
        }
        try {
            Authentication authenticated = authManager.authenticate(authentication);
            SecurityContextHolder.getContext().setAuthentication(authenticated);
            return 0;
        } catch (AuthenticationException e) {
            parseResult.commandSpec().commandLine().getErr().println(e.getMessage());
            return -3;
        }
    }

}

File: nekoshlink/shlink-security/src/main/kotlin/org/nekosoft/shlink/sec/cli/AuthenticationExecutionStrategy.kt
package org.nekosoft.shlink.sec.cli

import org.nekosoft.shlink.sec.ApiKeyAuthenticationToken
import org.springframework.boot.autoconfigure.condition.ConditionalOnNotWebApplication
import org.springframework.security.authentication.AnonymousAuthenticationToken
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.server.resource.BearerTokenAuthenticationToken
import org.springframework.stereotype.Component
import picocli.CommandLine.*
import java.io.File

@Component
@ConditionalOnNotWebApplication
class AuthenticationExecutionStrategy(
    private val authManager: AuthenticationManager
) : IExecutionStrategy {
    override fun execute(parseResult: ParseResult): Int {
        val username = parseResult.matchedOption("--usr")?.getValue<String>()
        val password = parseResult.matchedOption("--pwd")?.getValue<String>()
        val apiKey = parseResult.matchedOption("--api-key")?.getValue<String>()
        val accessTokenFile = parseResult.matchedOption("--access-token-file")?.getValue<File>()
        val authentication = if (accessTokenFile != null && accessTokenFile.canRead()) {
            BearerTokenAuthenticationToken(accessTokenFile.readText())
        } else if (!username.isNullOrBlank() && !password.isNullOrBlank()) {
            UsernamePasswordAuthenticationToken(username, password)
        } else if (!apiKey.isNullOrBlank()) {
            ApiKeyAuthenticationToken(apiKey)
        } else {
            val authorities = HashSet<GrantedAuthority>()
            authorities.add(SimpleGrantedAuthority("ROLE_Anyone"))
            AnonymousAuthenticationToken("NekoShlink", "anonymous", authorities)
        }
        return try {
            val authenticated = authManager.authenticate(authentication)
            SecurityContextHolder.getContext().authentication = authenticated
            0
        } catch (e: AuthenticationException) {
            parseResult.commandSpec().commandLine().err.println(e.message)
            -3
        }
    }
}